include/triton/core/tritonbackend.h

```diff
@@ -1206,6 +1227,28 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ModelInstanceState(
 TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ModelInstanceSetState(
     TRITONBACKEND_ModelInstance* instance, void* state);
 
+/// Report the memory usage of the model instnace that will be released on
+/// TRITONBACKEND_ModelInstanceFinalize. The backend may call this function
+/// within the lifecyle of the TRITONBACKEND_Model object (between
+/// TRITONBACKEND_ModelInstanceInitialize and
+/// TRITONBACKEND_ModelInstanceFinalize) to report the latest usage. To report
+/// the memory usage of the model, see TRITONBACKEND_ModelReportMemoryUsage.
+///
+/// \param instance The model instance.
+/// \param usage The list of buffer attributes that records the memory usage,
+/// each entry should record the total memory usage of a given memory type and
+/// id. For example, if the instance itself occupies 64 bytes on each of
+/// CUDA device 0 and CUDA device 1. Then 'usage' should have first two entries
+/// set, one has the buffer attributes of "type GPU, id 0, 64 bytes" and the
+/// other has "type GPU, id 1, 64 bytes". 'usage' is owned by the backend and
+/// may be released after the function returns.
+/// \param usage_size The number of entries in 'usage'.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceReportMemoryUsage(
```

Comment:
I was wondering why do we need two separate APIs to report the usage (ModelInstanceReportMemoryUsage and ModelReportMemoryUsage)? Is it because the CUPTI APIs track the memory per thread and the thread that loaded the model would be different than the thread that loads the model instance?

Reply:
How about having only a single model-level memory reporting that would account for the total memory usage? I understand that we can provide this breakdown because of the CUPTI thread level support but I think having a single total memory usage would be less confusing for the users. Also, it would create a lot of confusion specially when model sharing is enabled as you point out.

Reply:
After some thought - I agree that keeping it at model level with the expectation that all memory usage for all instances are combined there is simpler to specify. Initially I thought model instance made more sense but after a little more experience with the backendapi - I think it might be tricky to specify the correct behavior. We have to be explicit if model_size = sum(model_instances) + delta - since we don't have metrics at instance level (yet?) - maybe we defer this to future and leave it out for now.

Reply:
>but I think having a single total memory usage would be less confusing for the users.

What do you refer to for "users"?
- For user of Triton, they are somewhat far from the backend API, we can collect the memory usage as finest granularity and report the total usage for them, which is the current implementation at server frontend.
- For user of the backend API, I would say it is easier for them to manage the usage records. The issue with single model-level reporting is that they need to keep the summary and modify it as instances are being created / destroyed. Not to mention that the change of model instances may be in parallel in the future (i.e. RFE on concurrent instance loading) which further complicates the management. With this two-level reporting, the scope is constrained that model reports usage w.r.t. TritonModel and instance reports usage w.r.t. TritonModelInstance, and the usage record is not being shared (can even use it as local).

i.e. before
```
ModelState {
   ...
   usage_;
} 

ModelInit() {
  track and set "usage_"
}

InstanceInit() {
  declare lusage;
  ... track in lusage;
  get "usage_" from model state
  // thread-safety ..
  update "usage_"
  report model usage
}

InstanceFini() {
  // similar to above
}
```
After
```
ModelInit() {
  declare and track in lusage;
  report model usage;
}

InstanceInit() {
  declare and track in lusage;
  report instant usage;
}

InstanceFini() {
  // No extra handling, core already knows the instance usage
}
```

Reply:
So I understand: Is the suggestion that the core aggregates? So the "model" memory usage is "model + instance" as reported by the backend? As long as the semantics are clear and not left to the backend to decide - then ok. I want to make sure its clear what the backend implementation needs to be - i.e. model is aggregate or model is in addition to instances.

Reply:
> So the "model" memory usage is "model + instance" as reported by the backend?

Yes if `"model" memory usage` refers to what will be reported by statistic endpoint.

The backend API has mentioned what are the usage to be reported, let me know if it needs to be improved:
```
/// Report the memory usage of the model that will be released on
/// TRITONBACKEND_ModelFinalize. ...
TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ModelReportMemoryUsage
```
```
/// Report the memory usage of the model instnace that will be released on
/// TRITONBACKEND_ModelInstanceFinalize. ...
TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
TRITONBACKEND_ModelInstanceReportMemoryUsage(
```

Reply:
I think just giving the example / overall description that the total usage is model + model instances  (so backend doesn't double count by accident). Model Finalize could also release all instances etc. - so I think we should just be extra clear.

Reply:
Sorry for the delayed response. I think my only concern is that we may not able to bucket the model's memory usage to "model" and "model instance" buckets. For example, if we add support for memory usage tracking for activations (i.e., during the actual inference) we may not able "Untrack" it since it may not be returned to the system on model instance unloading.

On a related note, I was wondering how would this feature work with PyTorch and TF backend. Since they would not release the memory back to the system, we would probably untrack only a small portion (if any during model unloading). If another model ends up re-using the same memory pool, we wouldn't track any CUDA activity since it is using old allocations. Can this happen for TF and PyTorch? Is my understanding correct?

Reply:
For your first point, this is why the API is really generic that it is up to the backend to report the memory usage, if the backend is not going to release the memory on model instance unloading, then those memory should not be reported to the API.

For your second point, this is exactly what happened in TF and PyTorch, CUPTI collects no additional allocation and thus reports nothing. Ideally we should figure out what is the proper way for handling TF / PyTorch models and use corresponding implementation for those backends instead of the backend utils. One good example of this will be Python backend where it is not track-able by current implementation and the alternative is present that we can map the process-level usage to the model / instance usage.

After some thought, maybe the function name can be `XxxReportReleasableMemoryUsage` to show the intent of the API.

Reply:
I think the API naming is fine. It is probably more an issue with TF and PyTorch. We may be able to support the same API using Rapids RMM instead of Cupti.

Reply:
I think the API and metrics will be interpreted as memory_usage and releasable memory are the same. I think the expectation will be that the memory usage reflects the size of the model / instance and how many you could add / how much memory can be reclaimed. If there is a mismatch it may be confusing - for the cases that CUPTI works - my understanding is that they do match? (Overall usage and releasable usage per model?)